python/ray/serve/master.py

```diff
@@ -1,334 +1,528 @@
+
+        If the router does not already exist, it will be started.
+        """
+        try:
+            self.router = ray.util.get_actor(SERVE_ROUTER_NAME)
+        except ValueError:
+            logger.info(
+                "Starting router with name '{}'".format(SERVE_ROUTER_NAME))
+            self.router = async_retryable(router_class).options(
+                detached=True,
+                name=SERVE_ROUTER_NAME,
+                max_concurrency=ASYNC_CONCURRENCY,
+                max_reconstructions=ray.ray_constants.INFINITE_RECONSTRUCTION,
+            ).remote(**router_kwargs)
 
     def get_router(self):
-        assert self.router is not None, "Router not started yet."
+        """Returns a handle to the router managed by this actor."""
         return [self.router]
 
-    def start_http_proxy(self, host, port):
-        """Start the HTTP proxy on the given host:port.
+    def _get_or_start_http_proxy(self, host, port):
+        """Get the HTTP proxy belonging to this serve cluster.
 
-        On startup (or restart), the HTTP proxy will fetch its config via
-        get_http_proxy_config.
+        If the HTTP proxy does not already exist, it will be started.
         """
-        assert self.http_proxy is None, "HTTP proxy already started."
-        assert self.router is not None, (
-            "Router must be started before HTTP proxy.")
-        self.http_proxy = async_retryable(HTTPProxyActor).options(
-            max_concurrency=ASYNC_CONCURRENCY,
-            max_reconstructions=ray.ray_constants.INFINITE_RECONSTRUCTION,
-        ).remote(host, port)
-
-    async def get_http_proxy_config(self):
-        route_table = self.route_table.list_service(
-            include_methods=True, include_headless=False)
-        return route_table, self.get_router()
+        try:
+            self.http_proxy = ray.util.get_actor(SERVE_PROXY_NAME)
+        except ValueError:
+            logger.info(
+                "Starting HTTP proxy with name '{}'".format(SERVE_PROXY_NAME))
+            self.http_proxy = async_retryable(HTTPProxyActor).options(
+                detached=True,
+                name=SERVE_PROXY_NAME,
+                max_concurrency=ASYNC_CONCURRENCY,
+                max_reconstructions=ray.ray_constants.INFINITE_RECONSTRUCTION,
+            ).remote(host, port)
 
     def get_http_proxy(self):
-        assert self.http_proxy is not None, "HTTP proxy not started yet."
+        """Returns a handle to the HTTP proxy managed by this actor."""
         return [self.http_proxy]
 
-    def start_metric_monitor(self, gc_window_seconds):
-        assert self.metric_monitor is None, "Metric monitor already started."
-        self.metric_monitor = MetricMonitor.remote(gc_window_seconds)
-        # TODO(edoakes): this should be an actor method, not a separate task.
-        start_metric_monitor_loop.remote(self.metric_monitor)
-        self.metric_monitor.add_target.remote(self.router)
+    def get_http_proxy_config(self):
```

Comment:
just a thought, we should start adding some type annotations to make this easier
```python
RouteTable = Dict[str, Tuple[str, List[str]]
def get_http_proxy_config(self) -> Tuple[RouteTable, ActorHandle]:
    ...
```

Reply:
yeah I completely agree. that would be really helpful. I was going to do this but it was taking too long because I'm not familiar with python type annotations. Can do it in a follow up